.github/workflows/docker-publish.yml

```diff
@@ -1,30 +1,105 @@
-name: Publish Docker image (HEAD)
+name: Publish Docker images
 on:
   push:
-    branches: [main]
-    tags: ["v*.*.*"]
+    branches: main
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: "The tag of release"
+        required: true
 jobs:
-  push_to_registries:
-    name: Push Docker image to multiple registries
+  push_base_image:
+    name: Push base Docker image to multiple registries
     runs-on: ubuntu-latest
     steps:
       - name: Check out the repo
         uses: actions/checkout@v2
 
-      # TODO: Re-implement with v2 Docker GitHub Actions
-      # - name: Push to Docker Hub
-      #   uses: docker/build-push-action@v1
-      #   with:
-      #     username: ${{ secrets.DOCKER_USERNAME }}
-      #     password: ${{ secrets.DOCKER_PASSWORD }}
-      #     repository: grainlang/grain
-      #     tag_with_ref: true
+      - name: Parse tag
+        # This step converts Grain tags into standard semver, i.e. grain-v1.2.3 -> v1.2.3
+        id: vars
+        if: ${{ github.event.inputs.tag }}
+        run: GITHUB_TAG=${{ github.event.inputs.tag }}; echo ::set-output name=tag::${GITHUB_TAG#grain-}
 
-      - name: Push to GitHub Packages
-        uses: docker/build-push-action@v1
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v3
         with:
+          images: |
+            grainlang/grain
+            ghcr.io/grain-lang/grain
+          tags: |
+            type=ref,event=branch
+            type=semver,pattern={{version}},value=${{ steps.vars.outputs.tag }}
+            type=semver,pattern={{major}}.{{minor}},value=${{ steps.vars.outputs.tag }}
+
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+
+      - name: Login to Github Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build and push image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: Dockerfile
+          push: ${{ github.event_name != 'pull_request' }}
```

Comment:
If we only run this on `main`, why do we need to filter on `!= pull_request`?

Reply:
copypasta 🙈